text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+---
+feature name: fixing-type-unions
+start date: 2020-07-06
+rfc pr: 194
+related issue: 193
+---
+
+# Summary
+
+_Type unions_ is a feature of the TypeScript language that allows typing values as "one of several candidate types".
+This is achieved by specifying a `|`-delimited list of candidate types (e.g: `Foo | Bar`). While this feature is common
+in dynamic languages (**TypeScript**, **Python**, ...), it is almost always absent from statically typed languages
+(**Java**, **C#**, ...). In those languages, the impossibility to represent _type unions_ natively results in degraded
+developer experience due to the reduced fidelity of the generated typings.
+
+This RFC proposes to introduce new features in the _jsii_ compiler to support generation of additional code that will
+allow statically typed languages to benefit from the flexibility of type unions without having to endure the experience
+degradations they are currently exposed to.
+
+# Release Notes
+
+Use of **TypeScript** _type unions_ on any visible APIs (`public` or `protected` members of `export`ed types) within
+_jsii libraries_ must be named through the use of an exported type alias, and all those visible APIs must refer to the
+union using the alias:
+
+```ts
+export type ShinyUnion = Foo | Bar | Baz;
+
+export interface FancyProps {
+  readonly union: ShinyUnion;
+}
+```
+
+This aliasing offers multiple benefits for library authors and users alike:
+
+- _type unions_ are no longer repeated in multiple locations, so that new candidates can be added in a single location
+- better code can be generated to support _type union_ APIs in languages without native support for type unions (most
+  statically typed languages, such as **C#**, **Go**, **Java**, ...)
+- safer code can be authored, since _type unions_ no longer have to be frowned upon: all unsafe usage of
+  **TypeScript**'s `as any` to force a `Lazy` value can be replaced with a safer API using a _type union_ with
+  `IResolvable`
+
+An important think to keep in mind is that changing the type of a parameter to a _type union_ is a breaking change for
+those languages that do not natively support _type unions_. The `jsii-diff` tool can be used to automatically check that
+you incurred no such breakage on an API that was declared `@stable`. If you need to change a `@stable` API to support
+_type unions_ when it previously accepted only a single type, the easiest solution is to deprecate the current `@stable`
+API and introduce a new one that accepts or returns the _type union_.
+
+# Motivation
+
+## Existing use-cases
```

Comment:
I feel this section should go into the appendix and referenced from here so you can keep the motivation section more concise (basically just name the issues and reference an appendix which dives deeper). It buries the important part of the doc, which is the solution.

Reply:
I'd actually argue the important part of the doc is the problem statement. The solution is worthless if it does not address the problem! (Which was my issue before). That said - I'm all for moving the use-case characterization down, since it only provides context for the problem statement.